### PR TITLE
[examples/speech-recognition] Add SpecAugment to run_speech_recognition_seq2seq.py

### PR DESCRIPTION
**Pull request by bofenghuang**
# What does this PR do?

Hello 👋,

In this PR, I tried to add SpecAugment to run_speech_recognition_seq2seq.py training example since it's asked in https://github.com/huggingface/transformers/pull/21298.

I tried to not impact the training of other seq2seq models using this script. But I might still have missed something.

PS: Also removed useless argument `text_column` :)

Thanks in advance!

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

cc @ArthurZucker @sanchit-gandhi @sgugger 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 1]
```diff
@@ -113,6 +113,36 @@ class ModelArguments:
...
+    mask_feature_prob: float = field(
+        default=0.0,
+        metadata={
+            "help": "Percentage (between 0 and 1) of all feature vectors along the feature axis which will be masked. The masking procecure generates `mask_feature_prob*len(feature_axis)/mask_time_length` independent masks over the axis. If reasoning from the propability of each feature vector to be chosen as the start of the vector span to be masked, *mask_feature_prob* should be `prob_vector_start*mask_feature_length`. Note that overlap may decrease the actual percentage of masked vectors. This is only relevant if `apply_spec_augment is True`."
+        },
+    )
+    mask_feature_length: int = field(default=10, metadata={"help": "Length of vector span along the feature axis."})
+    mask_feature_min_masks: int = field(
+        default=0,
+        metadata={
```
Comment (sgugger) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:142]: This is making the example a bit too complex with all those new arguments. How about picking good defaults and put a comment telling the user they can customize them where you use it instead?

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 2]
```diff
@@ -367,6 +399,19 @@ def main():
...
+    if config.model_type == "whisper":
+        config.update(
+            {
+                "apply_spec_augment": model_args.apply_spec_augment,
+                "mask_time_prob": model_args.mask_time_prob,
+                "mask_time_length": model_args.mask_time_length,
+                "mask_time_min_masks": model_args.mask_time_min_masks,
+                "mask_feature_prob": model_args.mask_feature_prob,
+                "mask_feature_length": model_args.mask_feature_length,
+                "mask_feature_min_masks": model_args.mask_feature_min_masks,
```
Comment (sgugger) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:411]: I'd remove those to put good defaults and only do the update when `model_args.apply_spec_augment is True`. Then ut a comment telling the user to change the values to what they want.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! Adding 8 new arguments to the example for just one model is not consistent with our philosophy of keeping the example easy to understand and to customize, so I've left suggestions on how to add just one instead.

**Comment by bofenghuang** [comment]
Hi @sgugger,

Thanks for the review ! I've removed the SpecAugment arguments from `ModelArguments` and left some default values as asked. Also secured the attribute access in the line 447 to define `return_attention_mask` :)

**Review by sgugger — APPROVED** [review]
Thanks for iterating! Can you have a look and merge if you're happy @ArthurZucker or @sanchit-gandhi ?

**Review by ArthurZucker — APPROVED** [review]
LGTM, but @sanchit-gandhi has more experience with this!

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 3]
```diff
@@ -241,6 +246,9 @@ def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) ->
 
         batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt")
 
+        if self.return_attention_mask:
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:249]: This is potentially confusing since with Wav2Vec2 models, we'll return an attention mask when we call `feature_extractor.pad`:
```python
batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt")  # pad returns `input_values` and `attention_mask`
```
We can actually toggle whether this method returns an attention mask as well:
```python
batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt", return_attention_mask=False)  # pad returns `input_values` now
```

But with Whisper, we'll only get an attention mask if it's in `features`:
```python
batch["attention_mask"] = torch.LongTensor([feature["attention_mask"] for feature in features])
```

I can't think of a clean way of distinguishing the two other than perhaps a subtle name change:
```suggestion
        if self.forward_attention_mask:
```

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks @bofenghuang for finishing the Whisper SpecAug integration! Just a few comments regarding how we set the SpecAug params and one question about the argument `return_attention_mask` that we forward to the data collator!

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py (hunk 3 above)
Comment (bofenghuang) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:249]: Thanks for pointing this out and for the explication! But IMO wav2vec2 models are usually used in CTC architecture but not in run_speech_recognition_seq2seq.py no ?

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 4]
```diff
@@ -367,6 +375,23 @@ def main():
...
+                {
+                    "mask_time_prob": 0.05,
+                    "mask_time_length": 10,
+                    "mask_time_min_masks": 2,
+                    "mask_feature_prob": 0.0,
+                    "mask_feature_length": 10,
+                    "mask_feature_min_masks": 0,
+                }
+            )
+
```
[the remark this answers is not in the document]
Comment (bofenghuang) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:394]: Yeah! I was hesitating between not hard coding and giving some tips about these values! Totally agree with you! 

But should we remove the condition `getattr(config, "model_type", None) == "whisper"`? Not sure if it will confuse other models

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 5]
```diff
@@ -418,6 +443,12 @@ def main():
     text_column_name = data_args.text_column_name
     model_input_name = feature_extractor.model_input_names[0]
     do_lower_case = data_args.do_lower_case
+    # if SpecAugment is used for whisper models, return attention_mask to guide the mask along time axis
+    return_attention_mask = (
+        getattr(config, "model_type", None) == "whisper"
+        and getattr(config, "apply_spec_augment", False)
+        and getattr(config, "mask_time_prob", 0) > 0
+    )
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:451]: Thanks for the clarification! Indeed we only need the attention mask for time axis SpecAug 👍

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:394]: Cool! Let's keep in the `== "whisper"` condition then and remove the rest!

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py (hunk 3 above)
Comment (sanchit-gandhi) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:249]: We actually first added this script based on the results of this paper: https://arxiv.org/abs/2111.09296

Here, they create a sequence-to-sequence ASR model by initialising an encoder with a Wav2Vec2 model and adding a transformer decoder on top. We extended this further by initialising the encoder with Wav2Vec2 and the transformer decoder with a BART decoder: https://github.com/huggingface/transformers/tree/main/examples/pytorch/speech-recognition#warm-started-speech-encoder-decoder-model

Comment (bofenghuang) [reply examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py]: Thanks again for the explication! Done in https://github.com/huggingface/transformers/pull/21942/commits/19c7a24b57a0e05910326780257d058ee8965fbf

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 6]
```diff
@@ -367,6 +375,10 @@ def main():
 
     config.update({"forced_decoder_ids": model_args.forced_decoder_ids, "suppress_tokens": model_args.suppress_tokens})
 
+    # SpecAugment for whisper models
+    if getattr(config, "model_type", None) == "whisper":
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:379]: Here

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 7]
```diff
@@ -418,6 +430,12 @@ def main():
     text_column_name = data_args.text_column_name
     model_input_name = feature_extractor.model_input_names[0]
     do_lower_case = data_args.do_lower_case
+    # if SpecAugment is used for whisper models, return attention_mask to guide the mask along time axis
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:433]: Here

File: examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py [hunk 8]
```diff
@@ -241,6 +246,9 @@ def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) ->
 
         batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt")
 
+        if self.forward_attention_mask:
+            batch["attention_mask"] = torch.LongTensor([feature["attention_mask"] for feature in features])
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py:250]: Here

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for iterating @bofenghuang! From a modelling perspective, this looks good to me! Am super excited to hear how SpecAug changes your fine-tuning results; keen to stay in the loop with progress here!

My only question is really for @sgugger - are we good merging a fine-tuning script that has 3+ instances of Whisper-specific functionality? I've flagged them below for convenience. 

My understanding is that these scripts should be simple, barebones examples that generalise to all of the models in the library, that quickly showcase how one can use the library models for fine-tuning tasks.

If we continue by adding model-specific functionality for every new model that comes along, do we risk a barrage of `if/else` statements with model-specific logic for each new model that quickly convolutes the script and no longer generalises?

My feeling is that this is probably ok for Whisper since it's such an important model that has change the ASR field. But if it's against the `transformers` philosophy then maybe this is best kept as a script that builds on top of the repo example file? Keen to hear your thoughts here!

**Comment by sgugger** [comment]
I think it's okay to have the whisper-specific changes as they are in this PR. We have similar things in `run_translation` for some specific models as well.

**Comment by sanchit-gandhi** [comment]
Thanks for the clarification @sgugger and @bofenghuang for the contribution! Merging in that case 🤗
